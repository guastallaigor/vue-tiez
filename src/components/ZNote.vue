<template>
  <div class="z-note" :class="getClasses">
    <b v-if="type">{{ customLabel || type }}:</b>
    {{ content }}
    <div><slot/></div>
  </div>
</template>

<script>
export default {
  name: 'z-note',
  props: {
    content: {
      type: String,
      required: true,
      default: ''
    },
    type: {
      type: String,
      default: '',
      validator: (v) => (['', 'hint', 'alert', 'warning'].includes(v))
    },
    customLabel: {
      type: String,
      default: ''
    },
    dark: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    getClasses () {
      const { type, dark } = this

      return {
        hint: type === 'hint',
        alert: type === 'alert',
        warning: type === 'warning',
        dark
      }
    }
  }
}
</script>

<style lang="scss">
@mixin set-border-color-and-b($color) {
  border-color: $color;

  b {
    color: $color;
  }
}

.z-note {
  font-size: 14px;
  line-height: 1.8;
  color: rgb(0, 0, 0);
  padding: 16px 24px;
  border-radius: 4px;
  background: rgb(255, 255, 255);
  border-width: 1px;
  border-style: solid;
  border-color: rgb(221, 221, 221);
  border-image: initial;
  display: block;

  a {
    -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
    color: rgb(6, 125, 247);
    font-size: inherit;
    cursor: pointer;
    text-decoration: none;
  }

  b {
    text-transform: uppercase;
    font-weight: 500;
    color: rgb(0, 0, 0);
  }

  &.dark {
    color: rgb(255, 255, 255);
    background: rgb(0, 0, 0);

    b {
      color: rgb(255, 255, 255);
    }
  }

  &.hint {
    @include set-border-color-and-b(rgb(6, 125, 247));
  }

  &.alert {
    @include set-border-color-and-b(rgb(244, 129, 33));
  }

  &.warning {
    @include set-border-color-and-b(rgb(255, 0, 31));
  }
}
</style>
