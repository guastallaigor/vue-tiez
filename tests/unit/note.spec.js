import { shallowMount } from '@vue/test-utils'
import ZNote from '../../src/components/ZNote'

describe('When I create the ZNote component', () => {
  const createComponent = (propsData = {}, slot) => {
    return shallowMount(ZNote, {
      propsData,
      slots: {
        default: slot || 'ZNote'
      }
    })
  }

  it('should be a Vue instance', () => {
    const wrapper = createComponent({ content: 'Content' })
    expect(wrapper.find(ZNote).isVueInstance()).toBe(true)
  })

  it('should have a content and default type', () => {
    const wrapper = createComponent({ content: 'Content' })
    const note = wrapper.find('.z-note')
    expect(note.exists()).toBe(true)
    expect(note.classes().length).toBe(1)
    expect(note.classes()).toEqual(['z-note'])
  })

  it('should have a dark class', () => {
    const wrapper = createComponent({ content: 'Content', dark: true })
    const note = wrapper.find('.z-note')
    expect(note.exists()).toBe(true)
    expect(note.classes().length).toBe(2)
    expect(note.classes()).toEqual(['z-note', 'dark'])
  })

  it('should have a custom label', () => {
    const wrapper = createComponent({ content: 'Content', customLabel: 'CustomLabel' })
    const note = wrapper.find('.z-note')
    expect(note.exists()).toBe(true)
    expect(note.classes().length).toBe(1)
    expect(note.classes()).toEqual(['z-note'])
    const label = wrapper.find('.z-note > b')
    expect(label.exists()).toBe(true)
    expect(label.html()).toBe('<b>CustomLabel:</b>')
  })

  it('should have a type hint', () => {
    const wrapper = createComponent({ content: 'Content', type: 'hint' })
    const note = wrapper.find('.z-note')
    expect(note.exists()).toBe(true)
    expect(note.classes().length).toBe(2)
    expect(note.classes()).toEqual(['z-note', 'hint'])
    const label = wrapper.find('.z-note > b')
    expect(label.exists()).toBe(true)
    expect(label.html()).toBe('<b>hint:</b>')
  })

  it('should have a type alert', () => {
    const wrapper = createComponent({ content: 'Content', type: 'alert' })
    const note = wrapper.find('.z-note')
    expect(note.exists()).toBe(true)
    expect(note.classes().length).toBe(2)
    expect(note.classes()).toEqual(['z-note', 'alert'])
    const label = wrapper.find('.z-note > b')
    expect(label.exists()).toBe(true)
    expect(label.html()).toBe('<b>alert:</b>')
  })

  it('should have a type warning', () => {
    const wrapper = createComponent({ content: 'Content', type: 'warning' })
    const note = wrapper.find('.z-note')
    expect(note.exists()).toBe(true)
    expect(note.classes().length).toBe(2)
    expect(note.classes()).toEqual(['z-note', 'warning'])
    const label = wrapper.find('.z-note > b')
    expect(label.exists()).toBe(true)
    expect(label.html()).toBe('<b>warning:</b>')
  })

  it('should validate all the props type', () => {
    const wrapper = createComponent({ content: 'Content' })
    const type = wrapper.vm.$options.props.type
    expect(type.required).toBeFalsy()
    expect(type.type).toBe(String)
    expect(type.validator && type.validator('hint')).toBeTruthy()
    expect(type.validator && type.validator('alert')).toBeTruthy()
    expect(type.validator && type.validator('warning')).toBeTruthy()
    expect(type.validator && type.validator('')).toBeTruthy()
    const note = wrapper.find('.z-note')
    expect(note.exists()).toBe(true)
  })

  it('should have a slot instead of the content', () => {
    const wrapper = createComponent({ type: 'hint' }, '<span>test with slot</span>')
    const note = wrapper.find('.z-note')
    expect(note.exists()).toBe(true)
    expect(note.classes().length).toBe(2)
    expect(note.classes()).toEqual(['z-note', 'hint'])
    expect(note.html()).toBe(`<div class="z-note hint"><b>hint:</b>
  \n  <div><span>test with slot</span></div></div>`)
  })

  it('should match snapshot', () => {
    const wrapper = createComponent({
      content: 'Content',
      type: 'hint',
      customLabel: '',
      dark: false
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
