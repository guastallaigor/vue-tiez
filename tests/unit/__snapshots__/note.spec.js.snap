// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When I create the ZNote component should match snapshot 1`] = `
<div class="z-note hint"><b>hint:</b>
  Content
  <div>ZNote</div>
</div>
`;
